File: DFRobot_Alcohol.cpp
/*!
 * @file DFRobot_Alcohol.cpp
 * @brief Define the basic structure of the DFRobot_Alcohol class, the implementation of the basic methods
 * @copyright	Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license The MIT License (MIT)
 * @author [ZhixinLiu](zhixin.liu@dfrobot.com)
 * @version V1.0
 * @date 2020-09-09
 * @url https://github.com/DFRobot/DFRobot_Alcohol
 */
#include "DFRobot_Alcohol.h"

DFRobot_Alcohol::DFRobot_Alcohol(){}
DFRobot_Alcohol::~DFRobot_Alcohol(){}

void DFRobot_Alcohol::writeAlcohol(uint8_t reg, uint8_t pdata)
{
  if(uartI2CFlag == I2C_FLAG){
    sendBuf[0] = pdata;
    writeData(reg, sendBuf, 1);
  }else{
    if(pdata == MEASURE_MODE_AUTOMATIC){
      writeData(reg, cmdALOHA, 9);
    }else if(pdata == MEASURE_MODE_PASSIVE){
      writeData(reg, cmdPolling, 9);
    }
  }
}

void DFRobot_Alcohol::writeAlcohol_R(uint8_t reg, uint8_t pdata)
{
  if(uartI2CFlag == I2C_FLAG){
    sendBuf[0] = pdata;
    writeData(reg, sendBuf, 1);
  }
}

int16_t DFRobot_Alcohol::readAlcohol(uint8_t reg)
{
  if(uartI2CFlag == I2C_FLAG){
    if(-1 == readData(reg, readBuf, 2)){
      return -1;
    }else{
      return ((int16_t)readBuf[0] << 8) + readBuf[1];
    }
  }else{
    if(reg == AUTO_DATA_HIGE_REGISTER){
      readData(0, readBuf, DATA_LEN);
      if(readBuf[0] == 0xFF && readBuf[1] == 0x17){
        return ((int16_t)readBuf[4] << 8) + readBuf[5];
      }else{
        return -1;
      }
    }else if(reg == PASS_DATA_HIGE_REGISTER){
      writeData(0, cmdRead, DATA_LEN);
      readData (0, readBuf, DATA_LEN);
      if(readBuf[0] == 0xFF && readBuf[1] == 0x86){
        return ((int16_t)readBuf[6] << 8) + readBuf[7];
      }else{
        return -1;
      }
    }
  }
  return 0;
}

void DFRobot_Alcohol::setModes(uint8_t mode)
{
  if(mode == MEASURE_MODE_AUTOMATIC){
    writeAlcohol(MODE_REGISTER,  MEASURE_MODE_AUTOMATIC);
    delay(100);   _M_Flag = 0;
  }else if (mode == MEASURE_MODE_PASSIVE){
    writeAlcohol(MODE_REGISTER,  MEASURE_MODE_PASSIVE);
    delay(100);   _M_Flag = 1;
  }else {
    return;
  }
}

float DFRobot_Alcohol::readAlcoholData(uint8_t collectNum)
{
  static uint8_t i = 0, j = 0;
  if (collectNum > 0) {
    for(j = collectNum - 1;  j > 0; j--) {  alcoholData[j] = alcoholData[j-1]; }
    if(_M_Flag == 0){
      writeAlcohol_R(SET_PASSIVE_REGISTER,  AUTO_READ_DATA);      delay(100);
      alcoholData[0] = readAlcohol(AUTO_DATA_HIGE_REGISTER);
      if(alcoholData[0] == -1) {
        memset(alcoholData,0,collectNum);
        i = 0;
        j = 0;
        return -1;
      }
    }else if(_M_Flag == 1){
      writeAlcohol_R(SET_PASSIVE_REGISTER,  PASSIVE_READ_DATA);   delay(100);
      alcoholData[0] = readAlcohol(PASS_DATA_HIGE_REGISTER);
      if(alcoholData[0] == -1) {
        memset(alcoholData,0,collectNum);
        i = 0;
        j = 0;
        return -1;
      }
    }
    if(i < collectNum) i++;
    return getAverageNum(alcoholData, i);
  }else if(collectNum <= 0 || collectNum > 100){
   return -1;
  }
  return 0.0;
}

float DFRobot_Alcohol::getAverageNum(int bArray[], int iFilterLen) 
{
  uint8_t i = 0;
  unsigned long bTemp = 0;
  for(i = 0; i < iFilterLen; i++){
    bTemp += bArray[i];
  }
  return (float)(bTemp / iFilterLen) / 1000.0;
}

DFRobot_Alcohol_I2C::DFRobot_Alcohol_I2C(TwoWire *pWire, uint8_t addr)
{
  _pWire = pWire;
  this->_I2C_addr = addr;
  uartI2CFlag = I2C_FLAG;
}

bool DFRobot_Alcohol_I2C::begin()
{
  _pWire->begin();
  _pWire->beginTransmission(_I2C_addr);
  if(_pWire->endTransmission() == 0){
    return true;
  }else{
    return false;
  }
}

void DFRobot_Alcohol_I2C::writeData(uint8_t reg, uint8_t *data, uint8_t len)
{
  _pWire->beginTransmission(this->_I2C_addr);
  _pWire->write(reg);
  for(uint8_t i = 0; i < len; i++){
    _pWire->write(data[i]);
  }
  _pWire->endTransmission();
}

int16_t DFRobot_Alcohol_I2C::readData(uint8_t reg,uint8_t *data,uint8_t len)
{
  int i=0;
  _pWire->beginTransmission(this->_I2C_addr);
  _pWire->write(reg);
  if(_pWire->endTransmission() != 0){
    return -1;
  }
  _pWire->requestFrom((uint8_t)this->_I2C_addr,(uint8_t)len);
  while (_pWire->available()){
    data[i++]=_pWire->read();
  }
  return 0;
}

#ifdef ESP_PLATFORM
  DFRobot_Alcohol_UART::DFRobot_Alcohol_UART(HardwareSerial *hSerial, uint16_t Baud)
  {
    this->_serial = hSerial;
    this->_baud = Baud;
    uartI2CFlag = UART_FLAG;
  }
#else
  DFRobot_Alcohol_UART::DFRobot_Alcohol_UART(SoftwareSerial *sSerial, uint16_t Baud)
  {
    this->_serial = sSerial;
    this->_baud = Baud;
    uartI2CFlag = UART_FLAG;
  }
#endif

bool DFRobot_Alcohol_UART::begin()
{
  _serial->begin(this->_baud);
  this->writeData(0,cmdALOHA,DATA_LEN);
  this->readData (0,readBuf,DATA_LEN);
  if(readBuf[0] == 0xFF && readBuf[1] == 0x17){
    return true;
  }else{
    return false;
  }
}

void DFRobot_Alcohol_UART::writeData(uint8_t reg, uint8_t *data, uint8_t len)
{
  uint8_t i = 0;
  for(i = 0; i < len; i++){
    _serial->print((char)data[i]);
  }
}

int16_t DFRobot_Alcohol_UART::readData(uint8_t reg, uint8_t *data, uint8_t len)
{
  uint8_t i = 0;
  memset(readBuf,0,DATA_LEN);
  uint32_t nowtime = millis();
  while(millis() - nowtime < TIME_OUT){
    while(_serial->available() > 0){
      readBuf[i++] = _serial->read();
    }
  }
  return 0;
}

File: python/raspberrypi/DFRobot_Alcohol.py
# -*- coding: utf-8 -*
'''!
  @file DFRobot_Alcohol.py
  @brief DFRobot_Alcohol Class infrastructure, implementation of underlying methods
  @copyright Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license The MIT License (MIT)
  @author [ZhixinLiu](zhixin.liu@dfrobot.com)
  @version V1.0
  date 2020-09-09
  @url https://github.com/DFRobot/DFRobot_Alcohol
'''
import serial
import time
import smbus

ALCOHOL_ADDRESS_0         = 0x72
ALCOHOL_ADDRESS_1         = 0x73
ALCOHOL_ADDRESS_2         = 0x74
ALCOHOL_ADDRESS_3         = 0x75

## active  mode
MEASURE_MODE_AUTOMATIC    = 0x00
## passive mode
MEASURE_MODE_PASSIVE      = 0x01
## auto read alcohol data
AUTO_READ_DATA            = 0x00
## passive read alcohol data
PASSIVE_READ_DATA         = 0x01
## mode register
MODE_REGISTER             = 0x03
## read alcohol data register
SET_PASSIVE_REGISTER      = 0x04
## auto data high eight bits
AUTO_DATA_HIGE_REGISTER   = 0x09
## auto data Low  eight bits
AUTO_DATA_LOW_REGISTER    = 0x0A
## auto data high eight bits
PASS_DATA_HIGE_REGISTER   = 0x0B
## auto data Low  eight bits
PASS_DATA_LOW_REGISTER    = 0x0C
I2C_MODE                  = 0x01
UART_MODE                 = 0x02
ERROR                     = -1
  
class DFRobot_Alcohol(object):
  __m_flag   = 0                # mode flag
  __count    = 0                # acquisition count    
  __txbuf        = [0]          # i2c send buffer
  __alcoholdata  = [0]*101      # alcohol data
  __uart_i2c     =  0
  def __init__(self, bus, Baud):
    if bus != 0:
      self.i2cbus = smbus.SMBus(bus)
      self.__uart_i2c = I2C_MODE;
    else:
      self.ser = serial.Serial("/dev/ttyAMA0", baudrate=Baud,stopbits=1, timeout=0.5)
      self.__uart_i2c = UART_MODE;
      if self.ser.isOpen == False:
        self.ser.open()

  def set_mode(self, mode):
    '''!
      @brief set the mode
      @param MEASURE_MODE_AUTOMATIC  active mode
      @param MEASURE_MODE_PASSIVE    passive mode
    '''
    if mode ==  MEASURE_MODE_AUTOMATIC:
      self.__m_flag = 0
      self.__txbuf[0] = MEASURE_MODE_AUTOMATIC 
      self.write_reg(MODE_REGISTER, self.__txbuf)
    elif mode == MEASURE_MODE_PASSIVE:
      self.__m_flag = 1
      self.__txbuf[0] = MEASURE_MODE_PASSIVE
      self.write_reg(MODE_REGISTER, self.__txbuf)
    else:
      self.__m_flag = 2
      return

  def get_alcohol_data(self, collectnum):
    '''!
      @brief get the alcohol data, units of PPM
      @param collectnum Collect the number
      @return  alcohol concentration, (units PPM)
    '''
    if collectnum > 0:
      for num in range(collectnum, 1, -1):
        self.__alcoholdata[num-1] = self.__alcoholdata[num-2];
      if self.__m_flag == 0:
        self.__txbuf[0] = AUTO_READ_DATA
        self.write_reg(SET_PASSIVE_REGISTER, self.__txbuf)
        self.__alcoholdata[0] = self.get_alcohol(AUTO_DATA_HIGE_REGISTER)
        if self.__alcoholdata[0] == -1:
          for num in range (0, collectnum):
            self.__alcoholdata[num] = 0
          return -1
      elif self.__m_flag == 1:
        self.__txbuf[0] = PASSIVE_READ_DATA
        self.write_reg(SET_PASSIVE_REGISTER, self.__txbuf)
        self.__alcoholdata[0] = self.get_alcohol(PASS_DATA_HIGE_REGISTER)
        if self.__alcoholdata[0] == -1:
          for num in range (0, collectnum):
            self.__alcoholdata[num] = 0
          return -1
      if self.__count < collectnum:
        self.__count += 1
      return self.get_average_num(self.__alcoholdata, self.__count)
    elif (collectnum > 100) or (collectnum <= 0):
      return -1

  def get_average_num(self, barry, Len):
    '''!
      @brief get the average of the alcohol data, units of PPM
      @param barry alcohol data group
      @param Len The number of data
      @return  alcohol concentration, (units PPM)
    '''
    temp = 0
    for num in range (0, Len):
      temp += barry[num]
    return (temp / Len) / 1000.0

  def get_alcohol(self,reg):
    '''!
      @brief get the alcohol data, units of PPM
      @param reg register address
      @return  alcohol concentration, (units PPM)
    '''
    if self.__uart_i2c == I2C_MODE:
      rslt = self.read_reg(reg, 2)
      return ((rslt[0] << 8) + rslt[1])
    elif self.__uart_i2c == UART_MODE:
      rslt = self.read_reg(reg, 9)
      if rslt[0] == 0xff and rslt[1] == 0x17:
        number = rslt[4]<<8
        number += rslt[5]
        return number
      elif rslt[0] == 0xff and rslt[1] == 0x86:
        number = rslt[6]<<8
        number += rslt[7]
        return number
      else:
        return -1

class DFRobot_Alcohol_I2C(DFRobot_Alcohol): 
  def __init__(self, bus, addr):
    self.__addr = addr;
    super(DFRobot_Alcohol_I2C, self).__init__(bus,0)

  def write_reg(self, reg, data):
    while 1:
      try:
        self.i2cbus.write_i2c_block_data(self.__addr, reg, data)
        return
      except:
        print("please check connect!")
        time.sleep(1)

  def read_reg(self, reg, len):
    try:
      rslt = self.i2cbus.read_i2c_block_data(self.__addr, reg, len)
    except:
      rslt = -1
    return rslt

class DFRobot_Alcohol_UART(DFRobot_Alcohol): 
  __CmdALOHA   = [0xFF,0x01,0x78,0x40,0x00,0x00,0x00,0x00,0x47]
  __CmdPolling = [0xFF,0x01,0x78,0x41,0x00,0x00,0x00,0x00,0x46]
  __CmdRead    = [0xFF,0x01,0x86,0x00,0x00,0x00,0x00,0x00,0x79]


  def __init__(self, Baud):
    self.__Baud = Baud;
    super(DFRobot_Alcohol_UART, self).__init__(0,Baud)

  def write_reg(self, reg, data):
    if data[0] == MEASURE_MODE_AUTOMATIC and reg == MODE_REGISTER:
      self.ser.write(self.__CmdALOHA)
    elif data[0] == MEASURE_MODE_PASSIVE and reg == MODE_REGISTER:
      self.ser.write(self.__CmdPolling)
    elif data[0] == PASSIVE_READ_DATA    and reg == SET_PASSIVE_REGISTER:
      self.ser.write(self.__CmdRead)
    else:
      return;

  def read_reg(self, reg, len):
    timenow = time.time()
    recv = [0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
    i = 0
    while(time.time() - timenow) <= 1:
      count = self.ser.inWaiting()
      if count != 0:
        recv = self.ser.read(count)
        self.ser.flushInput()
        recv =[ord(c) for c in recv]
    return recv